### ISSUE TITLE
Presentational Tag Support

### ISSUE THREAD

**Issue by addihorowitz** [cdklabs/construct-hub-webapp#340]
![2021-08-16_07-28](https://user-images.githubusercontent.com/7221111/129580513-ec012211-1802-4941-94bd-e8d3af8d148e.png)

The current "tags" displayed on search result cards come from the list of keywords defined in the package.json of a package. However we want to allow displaying additional information such as whether the library is maintained by AWS or another trusted vendor. This could be connoted by an "official" tag and the conditions for what packages have this displayed should be configurable by the owners of the construct-hub instance.

This can be achieved some build time configuration allowing users to pass in an object describing a tag and its appearance, alongside a predicate function that dictates to what packages the tag is applied.

example `ch.config.ts`
```ts
export const config = {
  // order denotes priority of display, IE 'default' displays first if applied.
  tags: [{
    name: 'default',
    color: 'green',
    // predicate accepts package metadata and returns a boolean.
    predicate: (meta) => {
      return meta.author === 'AWS'
    }
  }]
}
```
This solution allows the frontend to handle displaying all of these "presentational" tags since this is essentially just computed state and there is nowhere to store this on the backend currently. Additionally, these tags must be searchable. Since all of the search logic is handled on the frontend currently, all we have to do is run every package in the catalog through all provided predicate functions, and add new keywords to the package list and use that new computed catalog to search against. If search logic is ever moved to the backend, or somewhere other than the react app, this will require computing and storing these somewhere that the search logic has access to.

**Comment by eladb** [comment]
@addihorowitz @MrArnoldPalmer what will prevent anyone from simply annotating their npm package with metadata that will cause their package to look as if it was certified/trusted/approved. npm metadata is completely controlled by the publisher of the module, so using only metadata to determine if a package is trusted is not enough.

I think we won't be able to get around including some backend state/support for this, perhaps piggyback on the mechanism we have for deny lists.

**Comment by gabewomble** [comment]
That's a concern I share as well. For the public construct hub, we can be more granular with what's allowed if needed. For example - with our official constructs, rather than only looking at the package's author name or email, we can validate the name's scope as well, which I don't think can be spoofed.

I think we already have this risk anyways as we rely on package keywords as tags right now, and they're also how we discover new constructs

**Comment by MrArnoldPalmer** [comment]
Even with some backend stateful data store I'm not sure how we solve this. The npm metadata is all we have really when it comes down to it.

I think most of these functions will be (pkg) => pkg.name === something, or whether the package exists in a certain npm scope. These seem to be the only reliable ways to assert ownership.

**Comment by eladb** [comment]
> I think most of these functions will be (pkg) => pkg.name === something, or whether the package exists in a certain npm scope. These seem to be the only reliable ways to assert ownership.

Yes, that is what I think as well, which is static state that's stored in the backend.

**Comment by MrArnoldPalmer** [comment]
Allowing these to run on the frontend still simplifies the implementation, preventing the need for running backfills when new/changed tags are added, assuming we are storing the results somewhere persistent.

Additionally, we could create a lambda@edge on the /catalog.json route that will run the tag functions and return the "new" version of the catalog with the computed tags applied and it will recompute it every so often.

**Comment by eladb** [comment]
> Additionally, we could create a lambda@edge on the /catalog.json route that will run the tag functions and return the "new" version of the catalog with the computed tags applied and it will recompute it every so often.

Not sure I follow - why can't we add these annotations when we generate `catalog.json` based on that static model (similar to how we exclude packages that match deny list rules).

I don't see how we can support this feature by only relying on package metadata that's controlled by publishers, even for internal hub.

**Comment by MrArnoldPalmer** [comment]
@eladb forget about the lambda@edge thing it's not a great solution.

My feeling is that these tags are useful for more than just the search cards, we likely will want to display "official" and other similar tags on the package details page. So in order for these to be available the tags themselves need to be in the package metadata file as well as the catalog.

We could implement this on the backend as part of metadata creation for newly ingested packages and the catalog update functionality, but it would require backfilling when  new tags are added by the owner.

>I don't see how we can support this feature by only relying on package metadata that's controlled by publishers, even for internal hub

I may be misinterpreting but what other option do we have? All of the content that we display in the app is computed from what is in the tarball. The things that we write to files on the backend is the stuff that is relatively expensive/slow to compute and comes with the downside of not being easily changed as we have to rewrite everything.

I'm under the impression that whether this logic is computed on the frontend or backend, the inputs can essentially be the same.

**Comment by eladb** [comment]
> I'm under the impression that whether this logic is computed on the frontend or backend, the inputs can essentially be the same.

You are right. This does not have to be in `catalog.json` or in tweaked package metadata. The frontend can simply download this static set of rules from the backend and render the UI accordingly.

**Comment by MrArnoldPalmer** [comment]
https://github.com/cdklabs/construct-hub-webapp/pull/446 is in progress and has the initial consumption of the backend config we added for custom tags. However since the new search implementation is in progress we want to wait to flesh that out before finishing this to avoid breaking existing functionality, like the current fake "official" tag click to search.
